libsolidity/analysis/PostTypeChecker.cpp

```diff
@@ -362,6 +365,77 @@ struct NoVariablesInInterfaceChecker: public PostTypeChecker::Checker
 	/// Flag indicating whether we are currently inside a StructDefinition.
 	int m_insideStruct = 0;
 };
+
+struct CircularContractDependencies: public PostTypeChecker::Checker
+{
+	CircularContractDependencies(ErrorReporter& _errorReporter):
+		Checker(_errorReporter)
+	{}
+
+	bool visit(ContractDefinition const& _contract) override
+	{
+		auto joined = boost::join(
+			_contract.annotation().creationCallGraph->get()->createdContracts,
+			_contract.annotation().deployedCallGraph->get()->createdContracts
+		);
```

Comment:
Yeah, we would lose it, but is that really a problem? It's only used for error reporting and if several expressions reference a contract within the same contract fixing one cycle would very likely also fix the other, and if not it will just show the error once you fixed the first one..

Reply:
Well, I spotted this at the beginning of the review and after going though it all, I think losing it is not such big a deal. The bigger problem for me is that there's no clear rule about which expression should eventually end up associated with the contract. I'm not even sure it's completely deterministic. `contractDependencies` gets updated in many places during the analysis and they are very loosely coupled.

Reply:
I am pretty sure it's deterministic as I had undeterministic problems before which no longer happen now..